File: backend/uploader/api/bosch_auth.py
import requests
from requests.auth import AuthBase
from requests import RequestException, HTTPError

from logger import logger
from pprint import pformat

def get_auth_token(client_id, client_secret, scope):
    url = 'https://access.bosch-iot-suite.com/token'

    try:
        response = requests.post(url, data={
            'grant_type': 'client_credentials',
            'client_id': client_id,
            'client_secret': client_secret,
            'scope': scope,
        })
        response.raise_for_status()
        response_json = response.json()
        return (response_json.token_type, response_json.access_token)
    except HTTPError as e:
            # log this first then rethrow
        logger.error('Failed to acquire bosch auth token due to the following HTTPError: {}'.format(e))
        raise e
    except RequestException as e:
        # log this first then rethrow
        logger.error('Failed to acquire bosch auth token due to the following RequestException: {}'.format(e))
        raise e


class BoschAuth(AuthBase):
    def __init__(self, client_id, client_secret, scope):
        self.cid = client_id
        self.csec = client_secret
        self.scope = scope

    def __call__(self, r):
        token_type, token = get_auth_token(self.cid, self.csec, self.scope)
        r.headers['Authorization'] = "{} {}".format(token_type, token)
        return r

File: backend/modules/uploader/api/boschAuth.js
const logger = require('../../logger').logger;
const axios = require('axios');

const getBoschAuthToken = async (clientId, clientSecret, scope) => {
    const url = 'https://access.bosch-iot-suite.com/token';

    try {
        const params = new URLSearchParams();
        params.append('grant_type', 'client_credentials');
        params.append('client_id', clientId);
        params.append('client_secret', clientSecret);
        params.append('scope', scope);

        const config = {
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
          },
        };

        const response = await axios.post(url, params, config);
        return `${response.data.token_type} ${response.data.access_token}`;
    } catch (e) {
        logger.log(`Failed to get Bosch Auth token due to the following error: ${JSON.stringify(e, null, 2)}`);
        throw e;
    }
};

module.exports = {
    getBoschAuthToken,
};
